internal/pkg/apply/apply.go

```diff
@@ -43,89 +48,294 @@ type Apply struct {
-	newhash, oknew := newAnnotation[inventory.HashAnnotation]
-	if okold && oknew && oldhash == newhash {
-		return obj, nil
+	// Retrieve inventory object with the InventoryId label.
+	selector := fmt.Sprintf("%s=%s", InventoryId, inventoryId)
+	listOptions := &metav1.ListOptions{LabelSelector: selector}
+	prevInventory := &unstructured.UnstructuredList{}
+	prevInventory.SetGroupVersionKind(schema.GroupVersionKind{
+		Kind:    "ConfigMapList",
+		Version: "v1",
+	})
+	err := a.DynamicClient.List(context.Background(), prevInventory,
+		director.GetNamespace(), listOptions)
+	if err != nil {
+		return fmt.Errorf("Error etrieving previous inventory objects: %#v", err)
+	}
+	// Store the resource ids from each inventory object.
+	for _, inv := range prevInventory.Items {
+		newInv := inventory.NewInventory()
+		err := newInv.LoadFromAnnotation(inv.GetAnnotations())
+		if err != nil {
+			return err
+		}
+		for resId, _ := range newInv.Current {
+			if a.PrevResources == nil {
+				a.PrevResources = make(map[resid.ResId]bool)
+			}
+			a.PrevResources[resId] = true
+		}
+		// Store the previous inventory object also.
+		a.PrevInventory = append(a.PrevInventory, inv)
 	}
 
-	return mergeContentAnnotation(u, obj)
+	return nil
 }
 
-func mergeContentAnnotation(newObj, oldObj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
-	newInv := inventory.NewInventory()
-	err := newInv.LoadFromAnnotation(newObj.GetAnnotations())
+func (a *Apply) getInventoryDirector(inv *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	director := inv.DeepCopy()
+	err := a.DynamicClient.Get(context.Background(),
+		types.NamespacedName{Namespace: inv.GetNamespace(), Name: inv.GetName()}, director)
 	if err != nil {
 		return nil, err
 	}
-	oldInv := inventory.NewInventory()
-	err = oldInv.LoadFromAnnotation(oldObj.GetAnnotations())
-	if err != nil {
-		return nil, err
+	return director, nil
+}
+
+func (a *Apply) createInventoryDirector(inv *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	director := &unstructured.Unstructured{}
+	director.SetGroupVersionKind(inv.GroupVersionKind())
+	director.SetNamespace(inv.GetNamespace())
+	director.SetName(inv.GetName())
+	inventoryId := createUniqueId()
+	annotations := map[string]string{InventoryId: inventoryId}
+	director.SetAnnotations(annotations)
+	if !a.DryRun {
+		err := a.DynamicClient.Create(context.Background(), director, &metav1.CreateOptions{})
+		if err != nil {
+			return nil, err
+		}
 	}
-	newInv.Previous.Merge(oldInv.Previous)
-	newInv.Previous.Merge(oldInv.Current)
+	fmt.Fprintf(a.Out, "[created] %s/%s\n", director.GetKind(), inv.GetName())
 
-	annotations := newObj.GetAnnotations()
-	err = newInv.UpdateAnnotations(annotations)
-	if err != nil {
-		return nil, err
+	return director, nil
+}
+
+func (a *Apply) prune() {
+	for _, inv := range a.PrevInventory {
+		fmt.Fprintf(a.Out, "[deleted:meta] %s/%s\n", inv.GetKind(), inv.GetName())
```

Comment:
The pruning only prints the message for deleting the inventory object. Can we add some explanation that it also deletes the objects that are only owned by this inventory object?